File: backend/internal/controller/settings/sitemap.go
package settings

import (
	"context"
	"fmt"
	"time"

	"blog/internal/dao"
	"blog/internal/model/entity"

	"github.com/gogf/gf/v2/frame/g"
)

// 生成sitemap.xml
type SitemapReq struct {
	g.Meta `path:"/sitemap.xml" method:"get" tags:"SEO" summary:"Generate sitemap.xml"`
}
type SitemapRes struct{}

// 生成robots.txt
type RobotsReq struct {
	g.Meta `path:"/robots.txt" method:"get" tags:"SEO" summary:"Generate robots.txt"`
}
type RobotsRes struct{}

func (c *ControllerV1) Sitemap(ctx context.Context, req *SitemapReq) (res *SitemapRes, err error) {
	r := g.RequestFromCtx(ctx)

	// 获取网站URL
	scheme := "https"
	if r.Header.Get("X-Forwarded-Proto") == "http" || r.URL.Scheme == "http" {
		scheme = "http"
	}
	siteUrl := fmt.Sprintf("%s://%s", scheme, r.Host)

	// 获取所有已发布的文章
	var articles []*entity.Article
	err = dao.Article.Ctx(ctx).
		Where(dao.Article.Columns().Status, "published").
		OrderDesc(dao.Article.Columns().Id).
		Scan(&articles)
	if err != nil {
		return nil, err
	}

	// 获取所有分类
	var categories []*entity.Category
	err = dao.Category.Ctx(ctx).Scan(&categories)
	if err != nil {
		return nil, err
	}

	// 获取所有标签
	var tags []*entity.Tag
	err = dao.Tag.Ctx(ctx).Scan(&tags)
	if err != nil {
		return nil, err
	}

	// 获取所有专题
	var topics []*entity.Topic
	err = dao.Topic.Ctx(ctx).Scan(&topics)
	if err != nil {
		return nil, err
	}

	// 生成sitemap XML
	xml := `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
`

	// 首页
	xml += fmt.Sprintf(`  <url>
    <loc>%s/</loc>
    <lastmod>%s</lastmod>
    <changefreq>daily</changefreq>
    <priority>1.0</priority>
  </url>
`, siteUrl, time.Now().Format("2006-01-02"))

	// 分类列表页
	xml += fmt.Sprintf(`  <url>
    <loc>%s/category</loc>
    <lastmod>%s</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.8</priority>
  </url>
`, siteUrl, time.Now().Format("2006-01-02"))

	// 标签列表页
	xml += fmt.Sprintf(`  <url>
    <loc>%s/tag</loc>
    <lastmod>%s</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.8</priority>
  </url>
`, siteUrl, time.Now().Format("2006-01-02"))

	// 专题列表页
	xml += fmt.Sprintf(`  <url>
    <loc>%s/topic</loc>
    <lastmod>%s</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.8</priority>
  </url>
`, siteUrl, time.Now().Format("2006-01-02"))

	// 关于页面
	xml += fmt.Sprintf(`  <url>
    <loc>%s/about</loc>
    <lastmod>%s</lastmod>
    <changefreq>monthly</changefreq>
    <priority>0.5</priority>
  </url>
`, siteUrl, time.Now().Format("2006-01-02"))

	// 文章详情页
	for _, article := range articles {
		updatedAt := article.UpdatedAt
		if updatedAt == nil {
			updatedAt = article.CreatedAt
		}
		lastmod := updatedAt.Format("2006-01-02")
		xml += fmt.Sprintf(`  <url>
    <loc>%s/article/%d</loc>
    <lastmod>%s</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.8</priority>
  </url>
`, siteUrl, article.Id, lastmod)
	}

	// 分类详情页
	for _, category := range categories {
		if category.Slug != "" {
			xml += fmt.Sprintf(`  <url>
    <loc>%s/category/%s</loc>
    <lastmod>%s</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.7</priority>
  </url>
`, siteUrl, category.Slug, time.Now().Format("2006-01-02"))
		}
	}

	// 标签详情页
	for _, tag := range tags {
		if tag.Slug != "" {
			xml += fmt.Sprintf(`  <url>
    <loc>%s/tag/%s</loc>
    <lastmod>%s</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.7</priority>
  </url>
`, siteUrl, tag.Slug, time.Now().Format("2006-01-02"))
		}
	}

	// 专题详情页
	for _, topic := range topics {
		xml += fmt.Sprintf(`  <url>
    <loc>%s/topic/%d</loc>
    <lastmod>%s</lastmod>
    <changefreq>weekly</changefreq>
    <priority>0.7</priority>
  </url>
`, siteUrl, topic.Id, time.Now().Format("2006-01-02"))
	}

	xml += `</urlset>`

	// 设置响应头
	r.Response.Header().Set("Content-Type", "application/xml; charset=utf-8")
	r.Response.Write(xml)

	return &SitemapRes{}, nil
}

func (c *ControllerV1) Robots(ctx context.Context, req *RobotsReq) (res *RobotsRes, err error) {
	r := g.RequestFromCtx(ctx)

	// 获取网站URL
	scheme := "https"
	if r.Header.Get("X-Forwarded-Proto") == "http" || r.URL.Scheme == "http" {
		scheme = "http"
	}
	siteUrl := fmt.Sprintf("%s://%s", scheme, r.Host)

	robots := fmt.Sprintf(`User-agent: *
Allow: /

# Sitemap
Sitemap: %s/sitemap.xml
`, siteUrl)

	r.Response.Header().Set("Content-Type", "text/plain; charset=utf-8")
	r.Response.Write(robots)

	return &RobotsRes{}, nil
}

File: ssr/server/api/sitemap.ts
// 生成 sitemap.xml
export default defineEventHandler(async (event) => {
  const config = useRuntimeConfig()
  const baseUrl = config.public.siteUrl || 'https://example.com'
  const apiBase = config.apiBase || 'http://localhost:8000/api'

  // 获取所有文章
  let articles: any[] = []
  try {
    const res = await $fetch<{ code: number; data: { list: any[] } }>(`${apiBase}/article/getList`, {
      method: 'POST',
      body: { page: 1, size: 1000, status: 'published' }
    })
    if (res.code === 0) {
      articles = res.data.list || []
    }
  } catch (error) {
    console.error('获取文章列表失败:', error)
  }

  // 获取所有分类
  let categories: any[] = []
  try {
    const res = await $fetch<{ code: number; data: { list: any[] } }>(`${apiBase}/category/getList`, {
      method: 'POST',
      body: { page: 1, size: 100 }
    })
    if (res.code === 0) {
      categories = res.data.list || []
    }
  } catch (error) {
    console.error('获取分类列表失败:', error)
  }

  // 获取所有标签
  let tags: any[] = []
  try {
    const res = await $fetch<{ code: number; data: { list: any[] } }>(`${apiBase}/tag/getList`, {
      method: 'POST',
      body: { page: 1, size: 100 }
    })
    if (res.code === 0) {
      tags = res.data.list || []
    }
  } catch (error) {
    console.error('获取标签列表失败:', error)
  }

  // 获取所有专题
  let topics: any[] = []
  try {
    const res = await $fetch<{ code: number; data: { list: any[] } }>(`${apiBase}/topic/getList`, {
      method: 'POST',
      body: { page: 1, size: 100 }
    })
    if (res.code === 0) {
      topics = res.data.list || []
    }
  } catch (error) {
    console.error('获取专题列表失败:', error)
  }

  // 生成 sitemap XML
  const urls: string[] = []

  // 静态页面
  const staticPages = [
    { loc: '/', priority: '1.0', changefreq: 'daily' },
    { loc: '/category', priority: '0.8', changefreq: 'weekly' },
    { loc: '/tag', priority: '0.8', changefreq: 'weekly' },
    { loc: '/topic', priority: '0.8', changefreq: 'weekly' },
    { loc: '/about', priority: '0.6', changefreq: 'monthly' }
  ]

  for (const page of staticPages) {
    urls.push(`
    <url>
      <loc>${baseUrl}${page.loc}</loc>
      <changefreq>${page.changefreq}</changefreq>
      <priority>${page.priority}</priority>
    </url>`)
  }

  // 文章页面
  for (const article of articles) {
    const lastmod = article.updatedAt || article.createdAt
    urls.push(`
    <url>
      <loc>${baseUrl}/article/${article.id}</loc>
      <lastmod>${new Date(lastmod).toISOString().split('T')[0]}</lastmod>
      <changefreq>weekly</changefreq>
      <priority>0.9</priority>
    </url>`)
  }

  // 分类页面
  for (const category of categories) {
    urls.push(`
    <url>
      <loc>${baseUrl}/category/${category.id}</loc>
      <changefreq>weekly</changefreq>
      <priority>0.7</priority>
    </url>`)
  }

  // 标签页面
  for (const tag of tags) {
    urls.push(`
    <url>
      <loc>${baseUrl}/tag/${tag.id}</loc>
      <changefreq>weekly</changefreq>
      <priority>0.7</priority>
    </url>`)
  }

  // 专题页面
  for (const topic of topics) {
    urls.push(`
    <url>
      <loc>${baseUrl}/topic/${topic.id}</loc>
      <changefreq>weekly</changefreq>
      <priority>0.8</priority>
    </url>`)
  }

  const sitemap = `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
${urls.join('')}
</urlset>`

  setResponseHeader(event, 'Content-Type', 'application/xml')
  return sitemap
})
